### PR TITLE
Search Command

### PR DESCRIPTION
**Pull request by splodingsocks**
I'm still in the middle of this, and I've got to run for this morning, but I wanted to get something up here for you to take a look at.

### CODE REVIEW

File: npm/flow-typed/jest.js.flow [hunk 1]
```diff
@@ -2,6 +2,7 @@ declare function beforeEach(cb: () => void): void;
 declare function describe(suiteName: string, cb: () => void): void;
 declare function expect(value: any): Object;
 declare function it(testName: string, cb: (done: () => void) => void): void;
+declare function pit(testName: string, cb: () => Promise<any>): void;
```
Comment (jeffmo) [inline npm/flow-typed/jest.js.flow]: Return type of the callback here should be `Promize<void>` I think

File: npm/src/lib/libDef.js [hunk 2]
```diff
@@ -346,3 +354,48 @@ export async function getShallowGHLibDefs(
...
+    repo: 'flow-typed',
+    path: `/definitions/${def.libDef.pkgNameVersionStr}`
+  })
+  .flatMap(identity)
+  .filter(notNull)
+  .flatMap(identity)
+  .map(i => i.name)
+  .filter(name => name.indexOf('flow-') == 0)
+  .map(name => {
+    let withoutPrefix = name.match(/flow-(.*)/)[1]
```
Comment (jeffmo) [inline npm/src/lib/libDef.js]: nit: might as well be `const`, eh?

(PS: "nit" === I don't feel strongly, feel free to object/ignore)

Comment (jeffmo) [inline npm/src/lib/libDef.js]: Also, surprised Flow is ok with this... .match() returns nullable (null if it doesn't find any matching patterns)

Comment (jeffmo) [inline npm/src/lib/libDef.js]: (Ahh, prob because flow doesn't have a libdef for RX)

**Comment by jeffmo** [comment]
Excite!

File: npm/src/lib/libDef.js (hunk 2 above)
Comment (splodingsocks) [inline npm/src/lib/libDef.js]: Good points!

**Comment by splodingsocks** [comment]
Oh, nice! It looks like flow 0.23 is out! And it broke our tests :)

**Comment by splodingsocks** [comment]
@jeffmo Thanks for reviewing even while on PTO :). I've added a search command, now! Should only be a few short steps to finish that command up, and then add an install command.

I still need to make the search command check to see what the current flow version is for the project. Do you think I should do that by running `flow version` and hope they have it in the path? Or should I look in the package.json for a `flow-bin`? Or should I do both? First looking at the package.json?

I think it'd be best to do both.

**Comment by splodingsocks** [comment]
K, I've done both. I think this could be merged now, unless you think it needs more automated tests?

File: npm/src/commands/search.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+function findFlowVersion(): string {
+  const nodeModulesBin = path.resolve('node_modules', '.bin')
+  const _path = `${nodeModulesBin}:${process.env.PATH}`
+
+  function fail() {
+    console.error('Couldn\'t identify your flow version. Please make sure ' +
+    'flow is installed and in your PATH, or flow-bin is installed ' +
+    'in your project.')
+    process.exit(-1)
```
Comment (jeffmo) [inline npm/src/commands/search.js]: Probably better to throw an `Error` than to `process.exit()`. That way anyone who potentially ends up using this elsewhere could, say, wrap the error message with some context (etc).

File: npm/src/commands/search.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
+// @flow
+import {getGHLibsAndFlowVersions, filterDefs, formatDefTable}
+  from "../lib/libDef.js";
+import type {LibDefWithFlow} from "../lib/libDef.js";
+import {child_process, path} from '../lib/node'
+
+const P = Promise;
```
Comment (jeffmo) [inline npm/src/commands/search.js]: Looks like you're not using `P` or `Promise`

File: npm/src/commands/search.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  const matches = output.match(/.*version (.*)\n$/)
+  if (!matches || !matches[1]) {
+    fail()
+  } else {
+    return matches[1]
+  }
+}
+
+export async function run(args: {}): Promise<number> {
+  if (!args._ || !args._.length > 1) {
```
Comment (jeffmo) [inline npm/src/commands/search.js]: `!args._.length > 1` will always be `false` -- it's the same as `(!args._.length) > 1`

File: npm/src/commands/search.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    console.error('Please provide a term for which to search!')
+    return 1;
+  }
+
+  const flowVersion = args.flowVersion || findFlowVersion()
```
Comment (jeffmo) [inline npm/src/commands/search.js]: I wonder if we should require the user to opt-in to searching for Flow? Doing it implicitly could easily end up with, say, a global version they have installed (even though maybe they're using something like `npm run flow` for all of their project-specific needs)?

Maybe something like `flow-typed --flowVersion=auto`?

File: npm/src/lib/libDef.js [hunk 7]
```diff
@@ -346,3 +356,80 @@ export async function getShallowGHLibDefs(
...
+      return stringToVersion(withoutPrefix)
+    } else {
+      return emptyVersion()
+    }
+  })
+  .toArray()
+  .toPromise()
+}
+
+export function getGHLibsAndFlowVersions(
```
Comment (jeffmo) [inline npm/src/lib/libDef.js]: Should probably call `_verifyCLIVersion()` in here first to be sure we're working with a version of the GH repo that we're compatible with

File: npm/src/commands/search.js (hunk 5 above)
Comment (splodingsocks) [inline npm/src/commands/search.js]: D'oh ⚡

File: npm/src/commands/search.js (hunk 6 above)
Comment (splodingsocks) [inline npm/src/commands/search.js]: What would you think of just omitting this functionality for now?

File: npm/src/lib/libDef.js (hunk 7 above)
Comment (splodingsocks) [inline npm/src/lib/libDef.js]: Awesome, thanks!

File: npm/src/commands/search.js (hunk 6 above)
Comment (jeffmo) [inline npm/src/commands/search.js]: That works too

**Comment by splodingsocks** [comment]
@jeffmo I think this is ready to merge now. Want to take a look?

File: npm/src/lib/__tests__/libDef-test.js [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+          pkgVersionStr: 'v0.3.x',
+          pkgNameVersionStr: 'mori_v0.3.x',
+          flowVersionStr: '>=v0.18.x',
+          flowVersion: { range: '>=', major: 0, minor: 18, patch: 'x' }
+        }
+      ]
+
+      const formatted = formatDefTable(fixture)
+      expect(formatted).toEqual(
+        "\nFound definitions:" + "\n" +
```
Comment (jeffmo) [inline npm/src/lib/__tests__/libDef-test.js]: ooo...

File: npm/src/lib/libDef.js [hunk 9]
```diff
@@ -388,3 +398,81 @@ export async function getShallowGHLibDefs(
   }
   return _libdefs
 }
+
+const identity = i => i
+
+function getGHFlowVersionsForDef(def: ShallowGHLibDef): Promise<Array<Version>> {
```
Comment (jeffmo) [inline npm/src/lib/libDef.js]: Need to call `_verifyCLIVersion()` in here too

(I'm going to send a commit to just add this in here myself -- just noting so I remember)

**Comment by splodingsocks** [comment]
Woo hoo!!

**Pull request merged**